Videos

keyvanamiri opened the pull request.

nicoya commented on app/src/main/AndroidManifest.xml:
```diff
@@ -70,10 +70,11 @@
             android:supportsPictureInPicture="true"
-            ></activity>
+            >
+        </activity>
```
Can just replace this with />

keyvanamiri replied:
Fixed.

nicoya commented on app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/media/MusicPlaybackService.java:
```diff
@@ -462,4 +493,74 @@ public void onDestroy() {
+    private class MediaSessionCallback extends MediaSession.Callback {
+
+        @Override
+        public boolean onMediaButtonEvent(Intent mediaButtonIntent) {
```
Do we need all the logging below? Seems pretty chatty.

keyvanamiri replied:
Fixed.

keyvanamiri committed:
Fixed pip media control button issues, made pip singleTask, added
NowPlayingCard for videos and added MediaSession activity for Music
and Videos

Pip play/pause control buttons weren't displayed which is now fixed. The
media players for both the videos and music service are now resposive to
media control buttons on the remote from anywhere outside the app. Added
NowPlayingCard for the currently active video item to the launcher.
Added MediaSession intent to be launched when clicking on
NowPlayingCard. Made PiP a singleTask (for now) so that no two videos
play at the same time.

nicoya commented on app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/media/MusicPlaybackService.java:
```diff
@@ -462,4 +493,74 @@ public void onDestroy() {
+        }
+
+        @Override
+        public void onSkipToNext() {
```
Just ignoring these? Btw, you can use these callbacks as a way to communicate between the front end (playback fragment/activity/whatever) and this service rather than using a bound service to make calls.

keyvanamiri replied:
The sample playback doesn't support fast-forward and rewind for now which I will add, although I'm not sure if with the current remote, we ever receive such events in the MediaSession callback.
Regarding the MediaSession callbacks, how can I pass custom objects (like a playlist) from the front-end to the service?

nicoya replied:
You can use the methods related to MediaSession.QueueItem for managing playlists in particular, or send custom messages via MediaSession#sendSessionEvent

nicoya commented on app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/media/VideoConsumptionExampleFragment.java:
```diff
@@ -60,40 +63,54 @@ protected void onRowChanged(PlaybackControlsRow row) {
         surface.getHolder().addCallback(new SurfaceHolder.Callback() {
             @Override
             public void surfaceCreated(SurfaceHolder holder) {
+                Log.d(TAG, "surfaceCreated called.");
```
Chatty logs. We fixed this issue, yeah? Don't need to keep these around.

keyvanamiri replied:
Fixed, Yes, setting display to null in surfaceDestroyed() fixed the crash when pressing play/pause button in gregarious mode.

nicoya commented on app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/media/VideoConsumptionExampleFragment.java:
```diff
@@ -60,40 +63,54 @@ protected void onRowChanged(PlaybackControlsRow row) {
-
         setBackgroundType(PlaybackOverlayFragment.BG_LIGHT);
         addPlaybackControlsRow();
+        Log.d(TAG, "onCreate finished.");
```
Super chatty.

keyvanamiri replied:
Fixed.

nicoya commented on app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/media/VideoExampleActivity.java:
```diff
@@ -22,6 +22,7 @@
 import android.os.Bundle;
 import android.support.v17.leanback.supportleanbackshowcase.R;
 import android.support.v4.os.BuildCompat;
+import android.util.Log;
```
Probably don't need this

keyvanamiri replied:
Fixed.

nicoya commented on app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/media/VideoMediaPlayerGlue.java:
```diff
@@ -91,16 +112,89 @@ public boolean abandonAudioFocus() {
+                              public void onResourceReady(
+                                      Bitmap resource,
+                                      GlideAnimation<? super Bitmap> glideAnimation) {
+                                  metaDataBuilder.putBitmap(MediaMetadata.METADATA_KEY_ALBUM_ART,
```
This pattern seems a bit sketchy. What happens if we want to load two bitmaps?

keyvanamiri replied:
So here I need to make sure that setMetaData is not called until the bitmap loading is complete from the URL. That's why I've put it inside the onResourceReady(). In case of loading two bitmaps, I think we should call a common method like setVideoSessionMetaData() which sets the MetaData on MediaSession only if both onResourceReady() from two bitmap loading are finished. We can define a flag for each bitmap and set it to true in its corresponding onResourceReady(). What do you think?

nicoya replied:
I dunno. It almost feels like using glide is a mismatch here. Is this method called on the main thread?

dakegu replied:
I would be surprised if Glide load bitmap in main thread.  This seems creating a glide task telling it to post to SimpleTarget on main thread after bitmap is loaded in background thread.

nicoya replied:
I meant the containing method. Obviously Glide will do its work off-thread.

nicoya commented on app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/media/VideoMediaPlayerGlue.java:
```diff
@@ -254,6 +383,37 @@ public void releaseResources() {
 
+    public void createMediaSessionIfNeeded() {
+        if (mVideoSession == null) {
+            mVideoSession = new MediaSession(this.getContext(), "VideoPlayer Session");
```
So we're creating a new MediaSession. Is this the right thing to do if we're recreating the activity due to a configuration change, or resuming playback from a now playing entry if the app was nuked in the background?

keyvanamiri replied:
Shouldn't  if (mVideoSession == null) check handle this case? A new media session is only created if it's already being released. Releasing a MediaSession happens in two place: fragment's onStop() and surfaceDestroyed().

The call to  createMediaSessionIfNeeded happens from two places: fragment's onStart() and prepareIfNeededAndPlay() (which is before playing a media item).  The latter is needed since it's possible that surfaceDestoryed is called without onStop() being called (If I'm playing a video in gregarious mode and press play/pause button and immediately click on NowPlayingCard for that video, surfaceDestroyed is called but before the fragment has a chance to stop I've returned to the activity and no onStart() call happens).

nicoya replied:
Well, I don't think we want to destroy and recreate the session on configuration change, so releasing it in onStop doesn't make sense.

My reading of the API is that the MediaSession should live in the playback service, and the UI should talk to it by way of a MediaController.
